libsolidity/analysis/TypeChecker.cpp

```diff
@@ -1498,6 +1498,30 @@ void TypeChecker::endVisit(BinaryOperation const& _operation)
 {
 	TypePointer const& leftType = type(_operation.leftExpression());
 	TypePointer const& rightType = type(_operation.rightExpression());
+
+	if (_operation.getOperator() == Token::SHL || _operation.getOperator() == Token::SAR)
+	{
+		auto const* rightIntegerType = dynamic_cast<IntegerType const*>(rightType);
+		if (!rightIntegerType)
+		{
+			auto const* rightRationalType = dynamic_cast<RationalNumberType const*>(rightType);
+			if (rightRationalType && !rightRationalType->isFractional())
+				rightIntegerType = rightRationalType->integerType();
+		}
+
+		if (!rightIntegerType || rightIntegerType->isSigned())
+			m_errorReporter.typeError(
+				_operation.location(),
+				"Operator " +
+				string(TokenTraits::toString(_operation.getOperator())) +
+				" not compatible with types " +
+				leftType->toString() +
+				" and " +
+				rightType->toString() +
+				". Right argument of shift operation should be unsigned."
```

Comment:
Can this be misleading when the `rightRationalType` is unsigned, but too big, so `rightIntegerType` becomes `nullptr`?